typing/parmatch.ml

```diff
@@ -786,10 +786,13 @@ let row_of_pat pat =
 
 (*
   Check whether the first column of env makes up a complete signature or
-  not.
+  not. We assume that the head patterns have already been simplified.
 *)
 
 let full_match closing env =  match env with
+| ({pat_desc = (Tpat_any|Tpat_var _|Tpat_alias _|Tpat_or _)},_) :: _ ->
+    assert false (* we assume we are working with simplified patterns *)
```

Comment:
We assume more than that.
You're allowed to assume the absence of variables, aliases and or-patterns on simplified matrices, but `Tpat_any` can still be present.
Here we can safely rule it out because we're called on the result on the `constrs` of `build_specialized_submatrices` and omega is not a discriminating constructor.

Reply:
Good point, I will fix the comment, thanks.

Yesterday I was vaguely considering the idea of having private types for sub-types of patterns (simplified patterns and heads/discriminants), but that could be a change for another day.

Reply:
I think I mentioned when we met a few months ago, but this summer I started a branch doing this. The thing is such a change also impacts matching (unless I messed something up, which is a possibility) and I at the time lacked the familiarity with matching to progress any further.
I still plan to revive that branch but yes, another day.